File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/model/gatt/client/pmd/model/DerivedAccData.kt
package com.polar.androidcommunications.api.ble.model.gatt.client.pmd.model

import com.polar.androidcommunications.api.ble.BleLogger
import com.polar.androidcommunications.api.ble.exceptions.PmdDataParseException
import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.PmdDataFrame
import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.PmdMeasurementType
import com.polar.androidcommunications.common.ble.TypeUtils

/**
 * Internal PMD model for derived ACC measurement data.
 *
 * One [DerivedAccSample] is produced per time window (e.g. 1 sample/s with 1000 ms window).
 *
 * @param activeMethods set of active method IDs (0-9, aligned to PolarDerivedMeasurementMethod.id)
 *   obtained from the recording settings [DERIVED_MEASUREMENT_METHOD].
 */
internal class DerivedAccData(val activeMethods: Set<Int> = emptySet()) {

    /**
     * One derived ACC output sample aligned to the end of its time window.
     *
     * @property timeStamp sample timestamp in nanoseconds
     * @property methodValues map from method ID → ordered list of output values.
     *   For spec-compliant type-15 frames: component methods (0-4) produce [x, y, z];
     *   scalar methods (5-9) produce [v].
     *   For current-firmware type-2 ACC frames: methods 0-3 produce [x, y, z];
     *   method 4 (Std) and methods 5-9 produce [v] (firmware scalar deviation).
     */
    data class DerivedAccSample internal constructor(
        val timeStamp: ULong,
        val methodValues: Map<Int, List<Int>>
    )

    val derivedSamples: MutableList<DerivedAccSample> = mutableListOf()

    companion object {

        /**
         * Methods 0-4 are all D→D (component-wise).
         * For 3D ACC: Downsample, Min, Max, Avg, Std each produce X, Y, Z values.
         * Used when parsing spec-compliant type-15 DERIVED_MEASUREMENT frames.
         */
        private val COMPONENT_METHODS_SPEC = setOf(0, 1, 2, 3, 4)

        /**
         * Method 4 (Std) deviates from spec and outputs one scalar per window.
         * Methods 5-9 are D→1 scalars per spec and firmware.
         */
        private val COMPONENT_METHODS_FIRMWARE = setOf(0, 1, 2, 3)

        /** Methods whose outputs are non-negative (unsigned-promoted per spec). */
        private val UNSIGNED_METHODS = setOf(4, 5, 6, 7, 8, 9)

        /**
         * Derived Data Measurement Frame header layout inside dataContent.
         *
         * Offset 0: Derived Data Frame Type (shall be 0)
         * Offset 1: Source Measurement Type
         * Offset 2: Source Measurement Frame Type
         * Offsets 3-4: Derived Measurement Method Bits (uint16 LE) — bitmask of active methods.
         * Offset 5+: Sample data
         *
         * This header is present ONLY for spec-compliant type-15 (DERIVED_MEASUREMENT) frames.
         * Current firmware stores derived output as type-2 ACC frames with no such header.
         */
        private const val DERIVED_FRAME_HEADER_SIZE = 5
        private const val DERIVED_FRAME_METHOD_BITS_OFFSET = 3

        fun parseDataFromDataFrame(frame: PmdDataFrame, activeMethods: Set<Int>): DerivedAccData {
            val result = DerivedAccData(activeMethods)
            if (frame.dataContent.isEmpty()) return result

            val isSpecDerivedFrame = frame.measurementType == PmdMeasurementType.DERIVED_MEASUREMENT

            val sampleDataBytes: ByteArray
            val effectiveMethods: List<Int>
            val componentMethods: Set<Int>

            if (isSpecDerivedFrame) {
                if (frame.dataContent.size < DERIVED_FRAME_HEADER_SIZE) return result
                val methodBits =
                    (frame.dataContent[DERIVED_FRAME_METHOD_BITS_OFFSET].toInt() and 0xFF) or
                    ((frame.dataContent[DERIVED_FRAME_METHOD_BITS_OFFSET + 1].toInt() and 0xFF) shl 8)
                effectiveMethods = (0..15).filter { (methodBits shr it) and 1 == 1 }
                sampleDataBytes = frame.dataContent.copyOfRange(
                    DERIVED_FRAME_HEADER_SIZE, frame.dataContent.size)
                componentMethods = COMPONENT_METHODS_SPEC
            } else {
                val sampleSizeInBytes = sampleSizeFromFrameType(frame.frameType)
                val sortedHint = activeMethods.sorted()
                var cumulative = 0
                val trimmed = mutableListOf<Int>()
                for (method in sortedHint) {
                    val vals = if (method in COMPONENT_METHODS_FIRMWARE) 3 else 1
                    val newTotal = cumulative + vals
                    if (newTotal * sampleSizeInBytes <= frame.dataContent.size) {
                        cumulative = newTotal
                        trimmed.add(method)
                    }
                }
                if (trimmed.size < sortedHint.size) {
                    BleLogger.w("DerivedAccData",
                        "parseDataFromDataFrame: hint $sortedHint trimmed to $trimmed " +
                        "to fit ${frame.dataContent.size} bytes (sampleSizeInBytes=$sampleSizeInBytes)")
                }
                effectiveMethods = trimmed
                sampleDataBytes = frame.dataContent
                componentMethods = COMPONENT_METHODS_FIRMWARE
            }

            if (effectiveMethods.isEmpty()) return result

            val sampleSizeInBytes = sampleSizeFromFrameType(frame.frameType)

            val valuesPerSample: Int = effectiveMethods.fold(0) { acc, m -> acc + if (m in componentMethods) 3 else 1 }
            val bytesPerSample: Int = valuesPerSample * sampleSizeInBytes
            if (bytesPerSample == 0) return result

            val samplesSize: Int = sampleDataBytes.size / bytesPerSample
            if (samplesSize <= 0) return result

            val timeStamps: List<ULong> = if (samplesSize == 1 || frame.previousTimeStamp == 0uL) {
                List(samplesSize) { frame.timeStamp }
            } else {
                val delta = (frame.timeStamp - frame.previousTimeStamp).toDouble() / samplesSize.toDouble()
                (0 until samplesSize).map { i ->
                    if (i == samplesSize - 1) frame.timeStamp
                    else (frame.previousTimeStamp + (delta * (i + 1)).toULong())
                }
            }

            // --- Parse samples -----------------------------------------------------------
            var offset = 0
            for (sampleIndex in 0 until samplesSize) {
                val methodValues = mutableMapOf<Int, List<Int>>()
                for (methodId in effectiveMethods) {
                    val isUnsigned = methodId in UNSIGNED_METHODS
                    if (methodId in componentMethods) {
                        val x = readValue(sampleDataBytes, offset, sampleSizeInBytes, isUnsigned); offset += sampleSizeInBytes
                        val y = readValue(sampleDataBytes, offset, sampleSizeInBytes, isUnsigned); offset += sampleSizeInBytes
                        val z = readValue(sampleDataBytes, offset, sampleSizeInBytes, isUnsigned); offset += sampleSizeInBytes
                        methodValues[methodId] = listOf(x, y, z)
                    } else {
                        val v = readValue(sampleDataBytes, offset, sampleSizeInBytes, unsigned = true); offset += sampleSizeInBytes
                        methodValues[methodId] = listOf(v)
                    }
                }
                result.derivedSamples.add(
                    DerivedAccSample(timeStamp = timeStamps[sampleIndex], methodValues = methodValues)
                )
            }
            return result
        }

        private fun sampleSizeFromFrameType(frameType: PmdDataFrame.PmdDataFrameType): Int =
            when (frameType) {
                PmdDataFrame.PmdDataFrameType.TYPE_0 -> 2
                PmdDataFrame.PmdDataFrameType.TYPE_1 -> 4
                PmdDataFrame.PmdDataFrameType.TYPE_2 -> 3
                else -> throw PmdDataParseException(
                    "DerivedAcc frame type $frameType is not supported"
                )
            }

        private fun readValue(data: ByteArray, offset: Int, length: Int, unsigned: Boolean): Int =
            if (unsigned) TypeUtils.convertArrayToUnsignedInt(data, offset, length).toInt()
            else TypeUtils.convertArrayToSignedInt(data, offset, length)
    }
}

File: sources/iOS/ios-communications/Sources/iOSCommunications/ble/api/model/gatt/client/pmd/model/DerivedAccData.swift
// Copyright © 2026 Polar Electro Oy. All rights reserved.

import Foundation

/// Internal PMD model for derived ACC data.
/// One `DerivedAccSample` is produced per time window (e.g. 1 Hz with a 1 000 ms window).
public class DerivedAccData {

    /// One derived ACC output sample aligned to the end of its time window.
    /// `methodValues`: method ID → values; component methods (0–3) give [x,y,z], scalars give [v].
    public struct DerivedAccSample {
        public let timeStamp: UInt64
        public let methodValues: [Int: [Int32]]
    }

    public let activeMethods: Set<Int>
    public var derivedSamples: [DerivedAccSample] = []

    public init(activeMethods: Set<Int> = []) {
        self.activeMethods = activeMethods
    }

    // MARK: - Parsing

    // Component methods for DERIVED_MEASUREMENT (type-15) frames per spec: 0–4 → [x,y,z].
    private static let componentMethodsDerivedFrame: Set<Int> = [0, 1, 2, 3, 4]

    // Component methods for current firmware (ACC type-2 frames): 0–3 → [x,y,z]; 4 and 5–9 are scalar.
    private static let componentMethodsFirmware: Set<Int> = [0, 1, 2, 3]

    // Methods whose outputs are unsigned-promoted.
    private static let unsignedMethods: Set<Int> = [4, 5, 6, 7, 8, 9]

    private static let derivedFrameHeaderSize = 4
    private static let derivedFrameMethodBitsOffset = 2

    static func parseDataFromDataFrame(frame: PmdDataFrame, activeMethods: Set<Int>) throws -> DerivedAccData {
        let result = DerivedAccData(activeMethods: activeMethods)
        guard !frame.dataContent.isEmpty else { return result }

        let sampleDataBytes: Data
        let effectiveMethods: [Int]
        let componentMethods: Set<Int>

        if frame.measurementType == .derivedMeasurement {
            guard frame.dataContent.count >= derivedFrameHeaderSize else { return result }
            let b3 = Int(frame.dataContent[derivedFrameMethodBitsOffset])
            let b4 = Int(frame.dataContent[derivedFrameMethodBitsOffset + 1])
            let methodBits = b3 | (b4 << 8)
            effectiveMethods = (0..<16).filter { (methodBits >> $0) & 1 == 1 }
            sampleDataBytes = frame.dataContent.subdata(in: derivedFrameHeaderSize..<frame.dataContent.count)
            componentMethods = componentMethodsDerivedFrame
        } else {
            let sampleSz = sampleSizeFromFrameType(frame.frameType)
            let sortedHint = activeMethods.sorted()

            let fullValuesPerSample = sortedHint.reduce(0) { acc, m in
                acc + (componentMethodsFirmware.contains(m) ? 3 : 1)
            }
            let fullBytesPerSample = fullValuesPerSample * sampleSz
            if fullBytesPerSample > 0 && frame.dataContent.count % fullBytesPerSample == 0 {
                effectiveMethods = sortedHint
            } else {
                var cumulative = 0
                var trimmed: [Int] = []
                for method in sortedHint {
                    let vals = componentMethodsFirmware.contains(method) ? 3 : 1
                    let newTotal = cumulative + vals
                    if newTotal * sampleSz <= frame.dataContent.count {
                        cumulative = newTotal
                        trimmed.append(method)
                    }
                }
                effectiveMethods = trimmed
            }
            sampleDataBytes = frame.dataContent
            componentMethods = componentMethodsFirmware
        }

        guard !effectiveMethods.isEmpty else { return result }

        let sampleSizeBytes = sampleSizeFromFrameType(frame.frameType)
        let valuesPerSample = effectiveMethods.reduce(0) { acc, m in acc + (componentMethods.contains(m) ? 3 : 1) }
        let bytesPerSample = valuesPerSample * sampleSizeBytes
        guard bytesPerSample > 0 else { return result }

        let samplesCount = sampleDataBytes.count / bytesPerSample
        guard samplesCount > 0 else { return result }

        let timeStamps = computeTimeStamps(
            frame: frame,
            samplesCount: samplesCount
        )

        var offset = 0
        for sampleIndex in 0..<samplesCount {
            var methodValues: [Int: [Int32]] = [:]
            for methodId in effectiveMethods {
                let isUnsigned = unsignedMethods.contains(methodId)
                if componentMethods.contains(methodId) {
                    let x = readValue(sampleDataBytes, offset: offset, length: sampleSizeBytes, unsigned: isUnsigned); offset += sampleSizeBytes
                    let y = readValue(sampleDataBytes, offset: offset, length: sampleSizeBytes, unsigned: isUnsigned); offset += sampleSizeBytes
                    let z = readValue(sampleDataBytes, offset: offset, length: sampleSizeBytes, unsigned: isUnsigned); offset += sampleSizeBytes
                    methodValues[methodId] = [x, y, z]
                } else {
                    let v = readValue(sampleDataBytes, offset: offset, length: sampleSizeBytes, unsigned: true); offset += sampleSizeBytes
                    methodValues[methodId] = [v]
                }
            }
            result.derivedSamples.append(DerivedAccSample(timeStamp: timeStamps[sampleIndex], methodValues: methodValues))
        }
        return result
    }

    private static func computeTimeStamps(frame: PmdDataFrame, samplesCount: Int) -> [UInt64] {
        if samplesCount == 1 || frame.previousTimeStamp == 0 {
            return Array(repeating: frame.timeStamp, count: samplesCount)
        }
        let delta = Double(frame.timeStamp - frame.previousTimeStamp) / Double(samplesCount)
        return (0..<samplesCount).map { i in
            i == samplesCount - 1
                ? frame.timeStamp
                : frame.previousTimeStamp + UInt64(delta * Double(i + 1))
        }
    }

    private static func sampleSizeFromFrameType(_ frameType: PmdDataFrameType) -> Int {
        switch frameType {
        case .type_0: return 2
        case .type_1: return 4
        case .type_2: return 3
        default: return 2 // fallback
        }
    }

    private static func readValue(_ data: Data, offset: Int, length: Int, unsigned: Bool) -> Int32 {
        guard offset + length <= data.count else { return 0 }
        if unsigned {
            var uVal: UInt32 = 0
            for i in 0..<min(length, 4) {
                uVal |= UInt32(data[offset + i]) << (i * 8)
            }
            return Int32(bitPattern: uVal)
        } else {
            return TypeUtils.convertArrayToSignedInt(data, offset: offset, size: length)
        }
    }
}

